sktime/utils/git_diff.py

```diff
@@ -78,3 +78,69 @@ class to get module string from, e.g., NaiveForecaster
     """
     module_str = get_module_from_class(cls)
     return is_module_changed(module_str)
+
+
+def get_changed_lines(file_path):
+    """Get changed or added lines from a file.
+
+    Parameters
+    ----------
+    file_path : str
+        path to file to get changed lines from
+
+    Returns
+    -------
+    list of str : changed lines
+    """
+    cmd = f"git diff remotes/origin/main -- {file_path}"
+
+    try:
+        # Run 'git diff' command to get the changes in the specified file
+        result = subprocess.check_output(cmd, shell=True, text=True)
+
+        # Extract the changed or new lines and return as a list of strings
+        changed_lines = [
+            line.strip() for line in result.split("\n") if line.startswith("+ ")
+        ]
+        # remove first character ('+') from each line
+        changed_lines = [line[1:] for line in changed_lines]
+
+        return changed_lines
+
+    except subprocess.CalledProcessError:
+        return []
+
+
+def get_packages_with_changed_specs():
+    """Get packages with changed or added specs.
+
+    Returns
+    -------
+    list of str : names of packages with changed or added specs
+    """
+    from packaging.requirements import Requirement
+
+    changed_lines = get_changed_lines("pyproject.toml")
+
+    packages = []
+    for line in changed_lines:
+        if line.find("'") > line.find('"') and line.find('"') != -1:
+            sep = '"'
+        elif line.find("'") == -1:
+            sep = '"'
+        else:
+            sep = "'"
+
+        req = line.split(sep)[1]
+
+        # deal with ; python_version >= "3.7" in requirements
+        if ";" in req:
+            req = req.split(";")[0]
+
+        pkg = Requirement(req).name
+        packages.append(pkg)
+
```

Comment:
Observed this. Can we use it somehow in this PR or somewhere else? Or is it already being used in cond1 check?

```pycon
>>>
>>> import sys
>>> sys.version_info
sys.version_info(major=3, minor=10, micro=12, releaselevel='final', serial=0)
>>>
>>> line_1 = """'numba<0.59,>=0.53; python_version < "3.12"'"""
>>> line_2 = """'numba<0.59,>=0.53; python_version < "3.9"'"""
>>>
>>> Requirement(line_1[1:-1]).marker.evaluate()
True
>>> Requirement(line_2[1:-1]).marker.evaluate()
False
>>>
```

Reply:
hm, interesting - I did not know that. Marker refers to the second part, right?

The logic currently is no as sharp - and I don't know how to get to sth like `line_1` easily to use this
You also have to think about different conditions, such as the entire line being added/changed, the marker being added/changed, and I don't see how we can dea with those without writing something much closer to a parser

If you have ideas though, happy to be shown some tricks

Reply:
What I was thinking is very optimistic, and may have other problems. What I was thinking is to do this:

1. start CI with a python 3.11+ job, which has `tomllib`.
2. read current `pyproject.toml` and that from `main`.
3. specifically compare which sets of dependency specifications vary, which will be available as dictionaries if I am not mistaken.
4. identify mismatched specifications.
5. identify names of packages from mismatches, and python requirements if any.
6. use it to find affected estimators, and affected environments if any.
7. trigger CI only for those environment-estimator combinations (related to #5719)

It's very different from the current PR I think, probably not worth considering. We can close this conversation.

Reply:
well, maybe worth tracking. My major question would be, how one woud get the poit of comparison. Get `pyproject.toml` from both branches and run through `tomllib`?

Your proposal seems more comprehensive, but also much more work...

Reply:
> Your proposal seems more comprehensive, but also much more work...

Agreed. That's why I'm very tempted to ignore as I don't know when to have time for this.

> Get pyproject.toml from both branches and run through tomllib?

Yes, exactly. I don't know the exact git command to fetch a single file from different branch as a separate file, but I'm sure we can find a way. At least by loading current contents once, checkout to other branch, read main contents, and back to old branch...some hacky trick like this is an easy way out, but may be there are better ways.

It's completely non-blocking for this PR.

Reply:
should we open an issue with this as an idea? Perhaps good first issue to replace either with de-novo or third party utility?